tests/src/common/test_case/app.rs

```diff
@@ -1,34 +1,209 @@
-use std::fmt::Debug;
+use std::{cell::RefCell, fmt::Debug, rc::Rc};
 
 use serde::Serialize;
 
 use finance::duration::Duration;
 use sdk::{
     cosmwasm_ext::{CosmosMsg, InterChainMsg},
-    cosmwasm_std::{Addr, BlockInfo, Coin as CwCoin, Empty, QuerierWrapper},
-    cw_multi_test::{AppResponse, Contract as CwContract, Executor},
-    testing::InterChainMsgReceiver,
+    cosmwasm_std::{
+        Addr, Api, Binary, BlockInfo, Coin as CwCoin, Empty, Querier, QuerierWrapper, Storage,
+        WasmMsg, WasmQuery,
+    },
+    cw_multi_test::{
+        AppResponse, Contract as CwContract, CosmosRouter, Executor, Wasm as WasmTrait, WasmKeeper,
+    },
+    testing::{CwApp, InterChainMsgReceiver},
 };
 
-use crate::common::{test_case::response::ResponseWithInterChainMsgs, AppExt as _, MockApp};
+use crate::common::{test_case::response::ResponseWithInterChainMsgs, AppExt as _};
+
+pub(crate) trait Wasm: WasmTrait<InterChainMsg, Empty> + Sized {
+    type CounterPart;
+
+    fn store_code(
+        app: &mut CwApp<Self>,
+        counter_part: &mut Self::CounterPart,
+        code: Box<dyn CwContract<InterChainMsg, Empty>>,
+    ) -> u64;
+}
+
+pub(crate) type DefaultWasm = WasmKeeper<InterChainMsg, Empty>;
+
+impl Wasm for DefaultWasm {
+    type CounterPart = ();
+
+    fn store_code(
+        app: &mut CwApp<Self>,
+        &mut (): &mut Self::CounterPart,
+        code: Box<dyn CwContract<InterChainMsg, Empty>>,
+    ) -> u64 {
+        app.store_code(code)
+    }
+}
+
+pub(crate) fn default_wasm() -> (DefaultWasm, ()) {
+    (DefaultWasm::new(), ())
+}
+
+pub(crate) enum Request<'r> {
+    Query(&'r WasmQuery),
+    Execute(&'r WasmMsg),
+    Sudo(&'r Addr),
+}
+
+pub(crate) enum Action {
+    Forward,
+    Error(anyhow::Error),
+}
+
+pub(crate) struct ConfigurableWasmBuilder<ActionSelectionF>
+where
+    ActionSelectionF: Fn(Request<'_>) -> Action,
+{
+    action_selection: ActionSelectionF,
+}
+
+impl<ActionSelectionF> ConfigurableWasmBuilder<ActionSelectionF>
+where
+    ActionSelectionF: Fn(Request<'_>) -> Action,
+{
+    pub const fn new(action_selection: ActionSelectionF) -> Self {
+        Self { action_selection }
+    }
+
+    pub fn build(
+        self,
+    ) -> (
+        ConfigurableWasm<ActionSelectionF>,
+        <ConfigurableWasm<ActionSelectionF> as Wasm>::CounterPart,
+    ) {
+        let inner: Rc<RefCell<DefaultWasm>> = Default::default();
+
+        (
+            ConfigurableWasm {
+                inner: inner.clone(),
+                action_selection: self.action_selection,
+            },
+            inner,
+        )
+    }
+}
+
+pub(crate) struct ConfigurableWasm<ActionSelectionF>
+where
+    ActionSelectionF: Fn(Request<'_>) -> Action,
+{
+    inner: Rc<RefCell<DefaultWasm>>,
```

Comment:
interior mutability needed??? could you elaborate on the rationale of defining `inner` of this type?

Reply:
The rationale is that I need to be able to call `store_code` from outside and CW's App stands in the way, not providing access to it. It's implementation is specialized for their own `WasmKeeper` structure and nothing else.
If you know a better solution, please let me know as this feedback would be useful for the future too!